packages/auth/src/access/types.ts

```diff
@@ -0,0 +1,178 @@
+  OutgoingPayment = 'outgoingPayment',
+  Quote = 'quote'
+}
+
+export enum Action {
+  Create = 'create',
+  Read = 'read',
+  List = 'list',
+  Complete = 'complete'
+}
+
+interface BaseAccessRequest {
+  actions: Action[]
+  locations?: string[]
+  identifier?: string
+  interval?: string
+}
+
+interface IncomingPaymentRequest extends BaseAccessRequest {
+  type: AccessType.IncomingPayment
+  limits?: IncomingPaymentLimit
+}
+
+interface OutgoingPaymentRequest extends BaseAccessRequest {
+  type: AccessType.OutgoingPayment
+  limits?: OutgoingPaymentLimit
+}
+
+interface AccountRequest extends BaseAccessRequest {
+  type: AccessType.Account
+  limits?: undefined
+}
+
+interface QuoteRequest extends BaseAccessRequest {
+  type: AccessType.Quote
+  limits?: undefined
+}
+
+export type AccessRequest =
+  | IncomingPaymentRequest
+  | OutgoingPaymentRequest
+  | AccountRequest
+  | QuoteRequest
+
+export function isAccessType(accessType: AccessType): accessType is AccessType {
+  return Object.values(AccessType).includes(accessType)
+}
+
+export function isAction(actions: Action[]): actions is Action[] {
+  if (typeof actions !== 'object') return false
+  for (const action of actions) {
+    if (!Object.values(Action).includes(action)) return false
+  }
+
+  return true
+}
+
+function isIncomingPaymentAccessRequest(
+  accessRequest: IncomingPaymentRequest
+): accessRequest is IncomingPaymentRequest {
+  return (
+    accessRequest.type === AccessType.IncomingPayment &&
+    isAction(accessRequest.actions) &&
+    (!accessRequest.limits || isIncomingPaymentLimit(accessRequest.limits))
+  )
+}
+
+function isOutgoingPaymentAccessRequest(
+  accessRequest: OutgoingPaymentRequest
+): accessRequest is OutgoingPaymentRequest {
+  return (
+    accessRequest.type === AccessType.OutgoingPayment &&
+    isAction(accessRequest.actions) &&
+    (!accessRequest.limits || isOutgoingPaymentLimit(accessRequest.limits))
+  )
+}
+
+function isAccountAccessRequest(
+  accessRequest: AccountRequest
+): accessRequest is AccountRequest {
+  return (
+    accessRequest.type === AccessType.Account &&
+    isAction(accessRequest.actions) &&
+    !accessRequest.limits
+  )
+}
+
+function isQuoteAccessRequest(
+  accessRequest: QuoteRequest
+): accessRequest is QuoteRequest {
+  return (
+    accessRequest.type === AccessType.Quote &&
+    isAction(accessRequest.actions) &&
+    !accessRequest.limits
+  )
+}
+
+export function isAccessRequest(
+  accessRequest: AccessRequest
+): accessRequest is AccessRequest {
+  return (
+    isIncomingPaymentAccessRequest(accessRequest as IncomingPaymentRequest) ||
+    isOutgoingPaymentAccessRequest(accessRequest as OutgoingPaymentRequest) ||
+    isAccountAccessRequest(accessRequest as AccountRequest) ||
+    isQuoteAccessRequest(accessRequest as QuoteRequest)
+  )
+}
+
+// value should hold bigint, serialized as string for requests & storage as jsonb field in postgres
+export interface PaymentAmount {
+  value: string
```

Comment:
Why not BigInt?

Reply:
Because the shape of the `limit` object can change depending on the `access.type` it is for, I stored as a JSON blob in postgres which doesn't play nicely with BigInt unless it is serialized somehow.

I figure any BigInt value being sent to the AS in a request or response would have to be serialized into a string anyway. The AS shouldn't be doing any arithmetic with them either, so I don't think there will be a need for the AS to de-serialize them either.